<a href="https://colab.research.google.com/github/AhmedELNajjar-dev/Depi-Final-Project/blob/main/MileStone(Data%20Cleaning)/Notebooks/MileStone(Cleaning_Data)_Finished.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Milestone 2 – Data Cleaning  
**Project:** Network Intrusion Detection System (CICIDS‑2017)  
**Team:** Bassem Gamal Mohamed – Dina Fawzy – Mohamed Farag – Pola Mokhtar – Ahmed Mamdouh – Mustafa Ibrahim  
Team Leader: Ahmed Mamdouh  
**Objective:** Remove inconsistencies, duplicates, and irrelevant columns to prepare a clean dataset for ML processing.  
Google Drive: https://drive.google.com/drive/folders/1wa5Y5heNduaFH-T2bIkQAkoXlhiy0QCe?usp=sharing

## Imports

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Make pandas tables easier to read
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)    # show all rows in Series/DataFrame
pd.set_option('display.max_columns', None) # show all columns

## Loading all csv files and reading them
> Head()  
> Info()  
> Describe()


In [4]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
file_paths = [
    '/content/drive/MyDrive/Depi Project/Data/TrafficLabelling/Monday-WorkingHours.pcap_ISCX.csv',
    '/content/drive/MyDrive/Depi Project/Data/TrafficLabelling/Tuesday-WorkingHours.pcap_ISCX.csv',
    '/content/drive/MyDrive/Depi Project/Data/TrafficLabelling/Wednesday-workingHours.pcap_ISCX.csv',
    '/content/drive/MyDrive/Depi Project/Data/TrafficLabelling/Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv',
    '/content/drive/MyDrive/Depi Project/Data/TrafficLabelling/Thursday-WorkingHours-Afternoon-Infilteration.pcap_ISCX.csv',
    '/content/drive/MyDrive/Depi Project/Data/TrafficLabelling/Friday-WorkingHours-Morning.pcap_ISCX.csv',
    '/content/drive/MyDrive/Depi Project/Data/TrafficLabelling/Friday-WorkingHours-Afternoon-PortScan.pcap_ISCX.csv',
    '/content/drive/MyDrive/Depi Project/Data/TrafficLabelling/Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv' ]

df_list = []
for path in file_paths:
    try:

        df = pd.read_csv(path, encoding="latin1", low_memory=False)
        df.columns = df.columns.str.strip()
        df_list.append(df)
    except FileNotFoundError:
        print(f"Warning: File not found at {path}. Skipping.")

df_full = pd.concat(df_list, ignore_index=True)

print(f"Original shape of all files combined: {df_full.shape}")

Original shape of all files combined: (3119345, 85)


In [6]:
df_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3119345 entries, 0 to 3119344
Data columns (total 85 columns):
 #   Column                       Dtype  
---  ------                       -----  
 0   Flow ID                      object 
 1   Source IP                    object 
 2   Source Port                  float64
 3   Destination IP               object 
 4   Destination Port             float64
 5   Protocol                     float64
 6   Timestamp                    object 
 7   Flow Duration                float64
 8   Total Fwd Packets            float64
 9   Total Backward Packets       float64
 10  Total Length of Fwd Packets  float64
 11  Total Length of Bwd Packets  float64
 12  Fwd Packet Length Max        float64
 13  Fwd Packet Length Min        float64
 14  Fwd Packet Length Mean       float64
 15  Fwd Packet Length Std        float64
 16  Bwd Packet Length Max        float64
 17  Bwd Packet Length Min        float64
 18  Bwd Packet Length Mean       float64
 19  

In [24]:
print("Descriptive statistics for numeric features:")
df_full.describe()

Descriptive statistics for numeric features:


Source Port  Destination Port      Protocol  \
count  2.827677e+06      2.827677e+06  2.827677e+06   
mean   4.113817e+04      8.060615e+03  9.884482e+00   
min    0.000000e+00      0.000000e+00  0.000000e+00   
25%    3.278400e+04      5.300000e+01  6.000000e+00   
50%    5.095000e+04      8.000000e+01  6.000000e+00   
75%    5.842000e+04      4.430000e+02  1.700000e+01   
max    6.553500e+04      6.553500e+04  1.700000e+01   
std    2.229169e+04      1.827364e+04  5.263201e+00   

                           Timestamp  Flow Duration  Total Fwd Packets  \
count                        2827677   2.827677e+06       2.827677e+06   
mean   2017-05-11 07:39:13.531523328   1.480170e+07       9.369491e+00   
min              2017-03-07 01:00:01  -1.300000e+01       1.000000e+00   
25%              2017-04-07 04:25:00   1.550000e+02       2.000000e+00   
50%              2017-05-07 10:48:00   3.134000e+04       2.000000e+00   
75%              2017-06-07 12:44:00   3.241964e+06       5.000000e+00   
max              2017-07-07 12:59:00   1.200000e+08       2.197590e+05   
std                              NaN   3.366846e+07       7.500791e+02   

       Total Backward Packets  Total Length of Fwd Packets  \
count            2.827677e+06                 2.827677e+06   
mean             1.040469e+01                 5.498907e+02   
min              0.000000e+00                 0.000000e+00   
25%              1.000000e+00                 1.200000e+01   
50%              2.000000e+00                 6.200000e+01   
75%              4.000000e+00                 1.880000e+02   
max              2.919220e+05                 1.290000e+07   
std              9.979288e+02                 9.998990e+03   

       Total Length of Bwd Packets  Fwd Packet Length Max  \
count                 2.827677e+06           2.827677e+06   
mean                  1.618017e+04           2.078189e+02   
min                   0.000000e+00           0.000000e+00   
25%                   2.000000e+00           6.000000e+00   
50%                   1.230000e+02           3.700000e+01   
75%                   4.840000e+02           8.100000e+01   
max                   6.554530e+08           2.482000e+04   
std                   2.264315e+06           7.175414e+02   

       Fwd Packet Length Min  Fwd Packet Length Mean  Fwd Packet Length Std  \
count           2.827677e+06            2.827677e+06           2.827677e+06   
mean            1.873052e+01            5.826028e+01           6.898295e+01   
min             0.000000e+00            0.000000e+00           0.000000e+00   
25%             0.000000e+00            6.000000e+00           0.000000e+00   
50%             2.000000e+00            3.400000e+01           0.000000e+00   
75%             3.600000e+01            5.000000e+01           2.616295e+01   
max             2.325000e+03            5.940857e+03           7.125597e+03   
std             6.035727e+01            1.861793e+02           2.813305e+02   

       Bwd Packet Length Max  Bwd Packet Length Min  Bwd Packet Length Mean  \
count           2.827677e+06           2.827677e+06            2.827677e+06   
mean            8.717917e+02           4.109211e+01            3.062790e+02   
min             0.000000e+00           0.000000e+00            0.000000e+00   
25%             2.000000e+00           0.000000e+00            2.000000e+00   
50%             7.900000e+01           0.000000e+00            7.200000e+01   
75%             2.820000e+02           7.700000e+01            1.810000e+02   
max             1.953000e+04           2.896000e+03            5.800500e+03   
std             1.947212e+03           6.888274e+01            6.055014e+02   

       Bwd Packet Length Std  Flow Bytes/s  Flow Packets/s  Flow IAT Mean  \
count           2.827677e+06  2.827677e+06    2.827677e+06   2.827677e+06   
mean            3.356893e+02  1.491624e+06    7.078873e+04   1.299857e+06   
min             0.000000e+00 -2.610000e+08   -2.000000e+06  -1.300000e+01   
25% 

Identifying Redundant Features for Next Milestone:

Looking back at the `.describe()` output, we can see several features where the `std` (standard deviation) is 0. This means the `min`, `max`, and `mean` are all identical, and the feature has zero variance.

Specifically, these are columns like `Bwd PSH Flags`, `Bwd URG Flags`, `Fwd Avg Bytes/Bulk`, and others. These features provide no predictive value.

these "redundant values" will be dropped in the **Milestone 3: Data Preprocessing** phase.

In [7]:
df_full.head()

Flow ID      Source IP  Source Port  \
0   192.168.10.5-8.254.250.126-49188-80-6  8.254.250.126         80.0   
1   192.168.10.5-8.254.250.126-49188-80-6  8.254.250.126         80.0   
2   192.168.10.5-8.254.250.126-49188-80-6  8.254.250.126         80.0   
3   192.168.10.5-8.254.250.126-49188-80-6  8.254.250.126         80.0   
4  192.168.10.14-8.253.185.121-49486-80-6  8.253.185.121         80.0   

  Destination IP  Destination Port  Protocol            Timestamp  \
0   192.168.10.5           49188.0       6.0  03/07/2017 08:55:58   
1   192.168.10.5           49188.0       6.0  03/07/2017 08:55:58   
2   192.168.10.5           49188.0       6.0  03/07/2017 08:55:58   
3   192.168.10.5           49188.0       6.0  03/07/2017 08:55:58   
4  192.168.10.14           49486.0       6.0  03/07/2017 08:56:22   

   Flow Duration  Total Fwd Packets  Total Backward Packets  \
0            4.0                2.0                     0.0   
1            1.0                2.0                     0.0   
2            1.0                2.0                     0.0   
3            1.0                2.0                     0.0   
4            3.0                2.0                     0.0   

   Total Length of Fwd Packets  Total Length of Bwd Packets  \
0                         12.0                          0.0   
1                         12.0                          0.0   
2                         12.0                          0.0   
3                         12.0                          0.0   
4                         12.0                          0.0   

   Fwd Packet Length Max  Fwd Packet Length Min  Fwd Packet Length Mean  \
0                    6.0                    6.0                     6.0   
1                    6.0                    6.0                     6.0   
2                    6.0                    6.0                     6.0   
3                    6.0                    6.0                     6.0   
4                    6.0                    6.0                     6.0   

   Fwd Packet Length Std  Bwd Packet Length Max  Bwd Packet Length Min  \
0                    0.0                    0.0                    0.0   
1                    0.0                    0.0                    0.0   
2                    0.0                    0.0                    0.0   
3                    0.0                    0.0                    0.0   
4                    0.0                    0.0                    0.0   

   Bwd Packet Length Mean  Bwd Packet Length Std  Flow Bytes/s  \
0                     0.0                    0.0     3000000.0   
1                     0.0                    0.0    12000000.0   
2                     0.0                    0.0    12000000.0   
3                     0.0                    0.0    12000000.0   
4                     0.0                    0.0     4000000.0   

   Flow Packets/s  Flow IAT Mean  Flow IAT Std  Flow IAT Max  Flow IAT Min  \
0    5.000000e+05            4.0           0.0           4.0           4.0   
1    2.000000e+06            1.0           0.0           1.0           1.0   
2    2.000000e+06            1.0           0.0           1.0           1.0   
3    2.000000e+06            1.0           0.0           1.0           1.0   
4    6.666667e+05            3.0           0.0           3.0           3.0   

   Fwd IAT Total  Fwd IAT Mean  Fwd IAT Std  Fwd IAT Max  Fwd IAT Min  \
0            4.0           4.0          0.0          4.0          4.0   
1            1.0           1.0          0.0          1.0          1.0   
2            1.0           1.0          0.0          1.0          1.0   
3            1.0           1.0          0.0          1.0          1.0   
4            3.0           3.0          0.0          3.0          3.0   

   Bwd IAT Total  Bwd IAT Mean  Bwd IAT Std  Bwd IAT Max  Bwd IAT Min  \
0            0.0           0.0          0.0          0.0          0.0   
1            0.0           0.0          0.0          0.0          0.0   
2 

In [8]:
rows_all_nan = df_full[df_full.isna().all(axis=1)]
print("Rows with all null:", rows_all_nan.shape[0])

Rows with all null: 288602


## Dropping all nulls rows in the dataframe

In [9]:
df_full = df_full.dropna(how='all')

In [10]:
df_full.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2830743 entries, 0 to 3119344
Data columns (total 85 columns):
 #   Column                       Dtype  
---  ------                       -----  
 0   Flow ID                      object 
 1   Source IP                    object 
 2   Source Port                  float64
 3   Destination IP               object 
 4   Destination Port             float64
 5   Protocol                     float64
 6   Timestamp                    object 
 7   Flow Duration                float64
 8   Total Fwd Packets            float64
 9   Total Backward Packets       float64
 10  Total Length of Fwd Packets  float64
 11  Total Length of Bwd Packets  float64
 12  Fwd Packet Length Max        float64
 13  Fwd Packet Length Min        float64
 14  Fwd Packet Length Mean       float64
 15  Fwd Packet Length Std        float64
 16  Bwd Packet Length Max        float64
 17  Bwd Packet Length Min        float64
 18  Bwd Packet Length Mean       float64
 19  Bwd P

In [11]:
df_full.isnull().sum()/len(df_full)*100

Flow ID                        0.000000
Source IP                      0.000000
Source Port                    0.000000
Destination IP                 0.000000
Destination Port               0.000000
Protocol                       0.000000
Timestamp                      0.000000
Flow Duration                  0.000000
Total Fwd Packets              0.000000
Total Backward Packets         0.000000
Total Length of Fwd Packets    0.000000
Total Length of Bwd Packets    0.000000
Fwd Packet Length Max          0.000000
Fwd Packet Length Min          0.000000
Fwd Packet Length Mean         0.000000
Fwd Packet Length Std          0.000000
Bwd Packet Length Max          0.000000
Bwd Packet Length Min          0.000000
Bwd Packet Length Mean         0.000000
Bwd Packet Length Std          0.000000
Flow Bytes/s                   0.047973
Flow Packets/s                 0.000000
Flow IAT Mean                  0.000000
Flow IAT Std                   0.000000
Flow IAT Max                   0.000000
Flow IAT Min                   0.000000
Fwd IAT Total                  0.000000
Fwd IAT Mean                   0.000000
Fwd IAT Std                    0.000000
Fwd IAT Max                    0.000000
Fwd IAT Min                    0.000000
Bwd IAT Total                  0.000000
Bwd IAT Mean                   0.000000
Bwd IAT Std                    0.000000
Bwd IAT Max                    0.000000
Bwd IAT Min                    0.000000
Fwd PSH Flags                  0.000000
Bwd PSH Flags                  0.000000
Fwd URG Flags                  0.000000
Bwd URG Flags                  0.000000
Fwd Header Length              0.000000
Bwd Header Length              0.000000
Fwd Packets/s                  0.000000
Bwd Packets/s                  0.000000
Min Packet Length              0.000000
Max Packet Length              0.000000
Packet Length Mean             0.000000
Packet Length Std              0.000000
Packet Length Variance         0.000000
FIN Flag Count                 0.000000
SYN Flag Count                 0.000000
RST Flag Count                 0.000000
PSH Flag Count                 0.000000
ACK Flag Count                 0.000000
URG Flag Count                 0.000000
CWE Flag Count                 0.000000
ECE Flag Count                 0.000000
Down/Up Ratio                  0.000000
Average Packet Size            0.000000
Avg Fwd Segment Size           0.000000
Avg Bwd Segment Size           0.000000
Fwd Header Length.1            0.000000
Fwd Avg Bytes/Bulk             0.000000
Fwd Avg Packets/Bulk           0.000000
Fwd Avg Bulk Rate              0.000000
Bwd Avg Bytes/Bulk             0.000000
Bwd Avg Packets/Bulk           0.000000
Bwd Avg Bulk Rate              0.000000
Subflow Fwd Packets            0.000000
Subflow Fwd Bytes              0.000000
Subflow Bwd Packets            0.000000
Subflow Bwd Bytes              0.000000
Init_Win_bytes_forward         0.000000
Init_Win_bytes_backward        0.000000
act_data_pkt_fwd               0.000000
min_seg_size_forward           0.000000
Active Mean                    0.000000
Active Std                     0.000000
Active Max                     0.000000
Active Min                     0.000000
Idle Mean                      0.000000
Idle Std                       0.000000
Idle Max                       0.000000
Idle Min                       0.000000
Label                          0.000000
dtype: float64

***0.047973 % are Nulls in Flow Bytes/s coloumn so we drop nulls for an easier approach since its a very small percentage***

In [12]:
df_full.dropna(inplace=True)

In [13]:
df_full.isna().sum().sum()

np.int64(0)

### Checking that Label coloumn has consistant data

In [26]:
df_full['Label'].value_counts()

Label
BENIGN                        2271122
DoS Hulk                       230123
PortScan                       158804
DDoS                           128025
DoS GoldenEye                   10293
FTP-Patator                      7935
SSH-Patator                      5897
DoS slowloris                    5796
DoS Slowhttptest                 5499
Bot                              1956
Web Attack – Brute Force         1507
Web Attack – XSS                  652
Infiltration                       36
Web Attack – Sql Injection         21
Heartbleed                         11
Name: count, dtype: int64

## Checking and handling Infinite Values



In [14]:
# Replace infinite values with NaN for easy dropping
df_full.replace([np.inf, -np.inf], np.nan, inplace=True)

# Check for any remaining NaN values after replacing inf
print("Number of NaN values after replacing infinite:", df_full.isnull().sum().sum())

# Drop rows with NaN values (which now include the original infinite values)
df_full.dropna(inplace=True)

print("Shape after removing infinite values:", df_full.shape)

Number of NaN values after replacing infinite: 3018
Shape after removing infinite values: (2827876, 85)


## Removing Non‑Informative Metadata Columns

The following column was removed:
**`Flow ID`**

**Reason for removal:**
- The field is **identifier**, not behavioral features of the network flow.  

In [15]:

drop_cols = ['Flow ID']

df_full.drop(columns=drop_cols, inplace=True)
print("Columns removed:", drop_cols)
print("New shape:", df_full.shape)

Columns removed: ['Flow ID']
New shape: (2827876, 84)


## Removing Duplicates

In [16]:
# Count all duplicated rows
duplicate_count = df_full.duplicated().sum()
print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 199


In [17]:
df_full.drop_duplicates(inplace=True)

In [18]:
df_full.head()

Source IP  Source Port Destination IP  Destination Port  Protocol  \
0  8.254.250.126         80.0   192.168.10.5           49188.0       6.0   
1  8.254.250.126         80.0   192.168.10.5           49188.0       6.0   
4  8.253.185.121         80.0  192.168.10.14           49486.0       6.0   
5  8.253.185.121         80.0  192.168.10.14           49486.0       6.0   
8   192.168.10.9       1031.0   192.168.10.3              88.0       6.0   

             Timestamp  Flow Duration  Total Fwd Packets  \
0  03/07/2017 08:55:58            4.0                2.0   
1  03/07/2017 08:55:58            1.0                2.0   
4  03/07/2017 08:56:22            3.0                2.0   
5  03/07/2017 08:56:22            1.0                2.0   
8  03/07/2017 08:56:38          609.0                7.0   

   Total Backward Packets  Total Length of Fwd Packets  \
0                     0.0                         12.0   
1                     0.0                         12.0   
4                     0.0                         12.0   
5                     0.0                         12.0   
8                     4.0                        484.0   

   Total Length of Bwd Packets  Fwd Packet Length Max  Fwd Packet Length Min  \
0                          0.0                    6.0                    6.0   
1                          0.0                    6.0                    6.0   
4                          0.0                    6.0                    6.0   
5                          0.0                    6.0                    6.0   
8                        414.0                  233.0                    0.0   

   Fwd Packet Length Mean  Fwd Packet Length Std  Bwd Packet Length Max  \
0                6.000000               0.000000                    0.0   
1                6.000000               0.000000                    0.0   
4                6.000000               0.000000                    0.0   
5                6.000000               0.000000                    0.0   
8               69.142857             111.967895                  207.0   

   Bwd Packet Length Min  Bwd Packet Length Mean  Bwd Packet Length Std  \
0                    0.0                     0.0               0.000000   
1                    0.0                     0.0               0.000000   
4                    0.0                     0.0               0.000000   
5                    0.0                     0.0               0.000000   
8                    0.0                   103.5             119.511506   

   Flow Bytes/s  Flow Packets/s  Flow IAT Mean  Flow IAT Std  Flow IAT Max  \
0  3.000000e+06    5.000000e+05            4.0      0.000000           4.0   
1  1.200000e+07    2.000000e+06            1.0      0.000000           1.0   
4  4.000000e+06    6.666667e+05            3.0      0.000000           3.0   
5  1.200000e+07    2.000000e+06            1.0      0.000000           1.0   
8  1.474548e+06    1.806240e+04           60.9    115.194956         381.0   

   Flow IAT Min  Fwd IAT Total  Fwd IAT Mean  Fwd IAT Std  Fwd IAT Max  \
0           4.0            4.0           4.0     0.000000          4.0   
1           1.0            1.0           1.0     0.000000          1.0   
4           3.0            3.0           3.0     0.000000          3.0   
5           1.0            1.0           1.0     0.000000          1.0   
8           2.0          609.0         101.5   177.089525        460.0   

   Fwd IAT Min  Bwd IAT Total  Bwd IAT Mean  Bwd IAT Std  Bwd IAT Max  \
0          4.0            0.0      0.000000     0.000000          0.0   
1          1.0            0.0      0.000000     0.000000          0.0   
4          3.0            0.0      0.000000     0.000000          0.0   
5          1.0            0.0      0.000000     0.000000          0.0   
8          2.0          467.0    155.666667   263.560872        460.0   

   Bwd IAT Min  Fwd PSH Flags  Bwd PSH Flags  Fwd URG Flags  Bwd URG Flags  \
0          0.0            

## Convert Timestamp to Datetime Objects

The `Timestamp` column is currently an object type. Converting it to datetime objects will enable time-series analysis and extraction of time-based features.

In [19]:
print("Converting Timestamp column to datetime...")
try:
    # Use format='mixed' to handle slight variations (like with or without seconds)
    # Explicitly set dayfirst=False or remove it
    df_full['Timestamp'] = pd.to_datetime(df_full['Timestamp'],
                                          format='mixed',
                                          dayfirst=False,  # <-- This is the fix
                                          errors='coerce')

    print("Timestamp conversion successful.")

    # Check the Dtype
    df_full.info(verbose=True)

except Exception as e:
    print(f"Error converting Timestamp: {e}")

Converting Timestamp column to datetime...
Timestamp conversion successful.
<class 'pandas.core.frame.DataFrame'>
Index: 2827677 entries, 0 to 3119344
Data columns (total 84 columns):
 #   Column                       Dtype         
---  ------                       -----         
 0   Source IP                    object        
 1   Source Port                  float64       
 2   Destination IP               object        
 3   Destination Port             float64       
 4   Protocol                     float64       
 5   Timestamp                    datetime64[ns]
 6   Flow Duration                float64       
 7   Total Fwd Packets            float64       
 8   Total Backward Packets       float64       
 9   Total Length of Fwd Packets  float64       
 10  Total Length of Bwd Packets  float64       
 11  Fwd Packet Length Max        float64       
 12  Fwd Packet Length Min        float64       
 13  Fwd Packet Length Mean       float64       
 14  Fwd Packet Length Std      

In [20]:
df_full.head()

Source IP  Source Port Destination IP  Destination Port  Protocol  \
0  8.254.250.126         80.0   192.168.10.5           49188.0       6.0   
1  8.254.250.126         80.0   192.168.10.5           49188.0       6.0   
4  8.253.185.121         80.0  192.168.10.14           49486.0       6.0   
5  8.253.185.121         80.0  192.168.10.14           49486.0       6.0   
8   192.168.10.9       1031.0   192.168.10.3              88.0       6.0   

            Timestamp  Flow Duration  Total Fwd Packets  \
0 2017-03-07 08:55:58            4.0                2.0   
1 2017-03-07 08:55:58            1.0                2.0   
4 2017-03-07 08:56:22            3.0                2.0   
5 2017-03-07 08:56:22            1.0                2.0   
8 2017-03-07 08:56:38          609.0                7.0   

   Total Backward Packets  Total Length of Fwd Packets  \
0                     0.0                         12.0   
1                     0.0                         12.0   
4                     0.0                         12.0   
5                     0.0                         12.0   
8                     4.0                        484.0   

   Total Length of Bwd Packets  Fwd Packet Length Max  Fwd Packet Length Min  \
0                          0.0                    6.0                    6.0   
1                          0.0                    6.0                    6.0   
4                          0.0                    6.0                    6.0   
5                          0.0                    6.0                    6.0   
8                        414.0                  233.0                    0.0   

   Fwd Packet Length Mean  Fwd Packet Length Std  Bwd Packet Length Max  \
0                6.000000               0.000000                    0.0   
1                6.000000               0.000000                    0.0   
4                6.000000               0.000000                    0.0   
5                6.000000               0.000000                    0.0   
8               69.142857             111.967895                  207.0   

   Bwd Packet Length Min  Bwd Packet Length Mean  Bwd Packet Length Std  \
0                    0.0                     0.0               0.000000   
1                    0.0                     0.0               0.000000   
4                    0.0                     0.0               0.000000   
5                    0.0                     0.0               0.000000   
8                    0.0                   103.5             119.511506   

   Flow Bytes/s  Flow Packets/s  Flow IAT Mean  Flow IAT Std  Flow IAT Max  \
0  3.000000e+06    5.000000e+05            4.0      0.000000           4.0   
1  1.200000e+07    2.000000e+06            1.0      0.000000           1.0   
4  4.000000e+06    6.666667e+05            3.0      0.000000           3.0   
5  1.200000e+07    2.000000e+06            1.0      0.000000           1.0   
8  1.474548e+06    1.806240e+04           60.9    115.194956         381.0   

   Flow IAT Min  Fwd IAT Total  Fwd IAT Mean  Fwd IAT Std  Fwd IAT Max  \
0           4.0            4.0           4.0     0.000000          4.0   
1           1.0            1.0           1.0     0.000000          1.0   
4           3.0            3.0           3.0     0.000000          3.0   
5           1.0            1.0           1.0     0.000000          1.0   
8           2.0          609.0         101.5   177.089525        460.0   

   Fwd IAT Min  Bwd IAT Total  Bwd IAT Mean  Bwd IAT Std  Bwd IAT Max  \
0          4.0            0.0      0.000000     0.000000          0.0   
1          1.0            0.0      0.000000     0.000000          0.0   
4          3.0            0.0      0.000000     0.000000          0.0   
5          1.0            0.0      0.000000     0.000000          0.0   
8          2.0          467.0    155.666667   263.560872        460.0   

   Bwd IAT Min  Fwd PSH Flags  Bwd PSH Flags  Fwd URG Flags  Bwd URG Flags  \
0          0.0            0.0   

In [28]:
df_full.describe()

Source Port  Destination Port      Protocol  \
count  2.827677e+06      2.827677e+06  2.827677e+06   
mean   4.113817e+04      8.060615e+03  9.884482e+00   
min    0.000000e+00      0.000000e+00  0.000000e+00   
25%    3.278400e+04      5.300000e+01  6.000000e+00   
50%    5.095000e+04      8.000000e+01  6.000000e+00   
75%    5.842000e+04      4.430000e+02  1.700000e+01   
max    6.553500e+04      6.553500e+04  1.700000e+01   
std    2.229169e+04      1.827364e+04  5.263201e+00   

                           Timestamp  Flow Duration  Total Fwd Packets  \
count                        2827677   2.827677e+06       2.827677e+06   
mean   2017-05-11 07:39:13.531523328   1.480170e+07       9.369491e+00   
min              2017-03-07 01:00:01  -1.300000e+01       1.000000e+00   
25%              2017-04-07 04:25:00   1.550000e+02       2.000000e+00   
50%              2017-05-07 10:48:00   3.134000e+04       2.000000e+00   
75%              2017-06-07 12:44:00   3.241964e+06       5.000000e+00   
max              2017-07-07 12:59:00   1.200000e+08       2.197590e+05   
std                              NaN   3.366846e+07       7.500791e+02   

       Total Backward Packets  Total Length of Fwd Packets  \
count            2.827677e+06                 2.827677e+06   
mean             1.040469e+01                 5.498907e+02   
min              0.000000e+00                 0.000000e+00   
25%              1.000000e+00                 1.200000e+01   
50%              2.000000e+00                 6.200000e+01   
75%              4.000000e+00                 1.880000e+02   
max              2.919220e+05                 1.290000e+07   
std              9.979288e+02                 9.998990e+03   

       Total Length of Bwd Packets  Fwd Packet Length Max  \
count                 2.827677e+06           2.827677e+06   
mean                  1.618017e+04           2.078189e+02   
min                   0.000000e+00           0.000000e+00   
25%                   2.000000e+00           6.000000e+00   
50%                   1.230000e+02           3.700000e+01   
75%                   4.840000e+02           8.100000e+01   
max                   6.554530e+08           2.482000e+04   
std                   2.264315e+06           7.175414e+02   

       Fwd Packet Length Min  Fwd Packet Length Mean  Fwd Packet Length Std  \
count           2.827677e+06            2.827677e+06           2.827677e+06   
mean            1.873052e+01            5.826028e+01           6.898295e+01   
min             0.000000e+00            0.000000e+00           0.000000e+00   
25%             0.000000e+00            6.000000e+00           0.000000e+00   
50%             2.000000e+00            3.400000e+01           0.000000e+00   
75%             3.600000e+01            5.000000e+01           2.616295e+01   
max             2.325000e+03            5.940857e+03           7.125597e+03   
std             6.035727e+01            1.861793e+02           2.813305e+02   

       Bwd Packet Length Max  Bwd Packet Length Min  Bwd Packet Length Mean  \
count           2.827677e+06           2.827677e+06            2.827677e+06   
mean            8.717917e+02           4.109211e+01            3.062790e+02   
min             0.000000e+00           0.000000e+00            0.000000e+00   
25%             2.000000e+00           0.000000e+00            2.000000e+00   
50%             7.900000e+01           0.000000e+00            7.200000e+01   
75%             2.820000e+02           7.700000e+01            1.810000e+02   
max             1.953000e+04           2.896000e+03            5.800500e+03   
std             1.947212e+03           6.888274e+01            6.055014e+02   

       Bwd Packet Length Std  Flow Bytes/s  Flow Packets/s  Flow IAT Mean  \
count           2.827677e+06  2.827677e+06    2.827677e+06   2.827677e+06   
mean            3.356893e+02  1.491624e+06    7.078873e+04   1.299857e+06   
min             0.000000e+00 -2.610000e+08   -2.000000e+06  -1.300000e+01   
25% 

### Identifying and Handling Impossible Values

The `.describe()` output revealed impossible negative values in time-based columns like `Flow Duration` (min: -13) and `Flow IAT Mean` (min: -13). A time duration cannot be negative, as this would imply a flow ended before it started.

[cite_start]These are clearly "unreliable values" [cite: 174, 184] and data errors, likely from the capture or calculation process. We must remove these rows to ensure data quality. We will **not** remove rows with a value of 0, as a zero duration is a valid measurement for a single-packet flow.

In [29]:
print(f"Original shape before removing negative durations: {df_full.shape}")

negative_rows = df_full[
    (df_full['Flow Duration'] < 0) |
    (df_full['Flow IAT Mean'] < 0)
].shape[0]

print(f"Found {negative_rows} rows with negative time values.")

df_full = df_full[df_full['Flow Duration'] >= 0]
df_full = df_full[df_full['Flow IAT Mean'] >= 0]

print(f"New shape after removing negative time values: {df_full.shape}")

Original shape before removing negative durations: (2827677, 84)
Found 115 rows with negative time values.
New shape after removing negative time values: (2827562, 84)


## Optimize Data Types

Based on the paper's methodology, we will optimize the data types to reduce memory usage. This involves downcasting integer columns to smaller integer types while keeping float columns as `float64`.

In [21]:
# 6. Optimize Data Types (Revised Method)
print("\nOptimizing data types (Revised Method)...")

# Check memory usage before
mem_before = df_full.memory_usage(deep=True).sum() / (1024**2)
print(f"Memory usage before downcasting: {mem_before:.2f} MB")

# Get all float columns
float_cols = df_full.select_dtypes(include=['float64']).columns
downcasted_cols = []

for col in float_cols:
    try:
        # Check if all values in the column are whole numbers
        # We do this by checking if the fractional part of all numbers is 0
        if (np.modf(df_full[col])[0] == 0).all():
            # If it's all whole numbers, downcast it to integer
            df_full[col] = pd.to_numeric(df_full[col], downcast='integer')
            downcasted_cols.append(col)
    except TypeError:
        # This can happen if there's an unexpected value; we'll just skip it
        print(f"Could not check column {col}, skipping.")

print(f"Successfully downcasted {len(downcasted_cols)} integer-like columns.")
print("True float columns (with decimals) were maintained.")

# Check memory usage after
mem_after = df_full.memory_usage(deep=True).sum() / (1024**2)
print(f"Memory usage after downcasting: {mem_after:.2f} MB")


Optimizing data types (Revised Method)...
Memory usage before downcasting: 2248.83 MB


/usr/local/lib/python3.12/dist-packages/pandas/core/dtypes/cast.py:378: RuntimeWarning: invalid value encountered in cast
  new_result = trans(result).astype(dtype)


Successfully downcasted 56 integer-like columns.
True float columns (with decimals) were maintained.
Memory usage after downcasting: 1474.88 MB


In [22]:
df_full.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2827677 entries, 0 to 3119344
Data columns (total 84 columns):
 #   Column                       Dtype         
---  ------                       -----         
 0   Source IP                    object        
 1   Source Port                  int32         
 2   Destination IP               object        
 3   Destination Port             int32         
 4   Protocol                     int8          
 5   Timestamp                    datetime64[ns]
 6   Flow Duration                int32         
 7   Total Fwd Packets            int32         
 8   Total Backward Packets       int32         
 9   Total Length of Fwd Packets  int32         
 10  Total Length of Bwd Packets  int32         
 11  Fwd Packet Length Max        int16         
 12  Fwd Packet Length Min        int16         
 13  Fwd Packet Length Mean       float64       
 14  Fwd Packet Length Std        float64       
 15  Bwd Packet Length Max        int16         
 16  Bwd P

## Saving Final Cleaned File

In [23]:
# Specify the path and filename for the output CSV file
output_path = "/content/drive/MyDrive/Depi Project/Cleaned Data/Cleaned_Data.csv"

# Save the DataFrame to CSV
df_full.to_csv(output_path, index=False)

print(f"Cleaned data saved to {output_path}")

Cleaned data saved to /content/drive/MyDrive/Depi Project/Cleaned Data/Cleaned_Data.csv
